### PR TITLE
feat: support CI/CD job token scope api

### LINKED ISSUE #571: Manage CI/CD job token scope allowlist
It would be great to be able to use GitLabForm to configure the allowlist of projects that are can use a project's `CI_JOB_TOKEN`.

https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html#configure-cicd-job-token-access

The API for this is https://docs.gitlab.com/ee/api/project_job_token_scopes.html

### PR DESCRIPTION
**Pull request by TimKnight-DWP**
closes #571

### CODE REVIEW

**Comment by TimKnight-DWP** [comment]
Resolves https://github.com/gitlabform/gitlabform/issues/549

**Review by amimas — COMMENTED** [review]
Looking forward to this feature @TimKnight-DWP . On my mobile and can't finish the review. Here's a few comments on what I looked at so far .

File: gitlabform/processors/project/job_token_scope_processor.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        )
+
+    def _process_groups(
+        self,
+        job_token_scope: ProjectJobTokenScope,
+        groups_allowlist: List,
+    ):
+        if not groups_allowlist:
+            warning(
+                "Process will remove existing groups from allowlist, as none set in configuration"
```
[the remark this answers is not in the document]
Comment (TimKnight-DWP) [reply gitlabform/processors/project/job_token_scope_processor.py:88]: @amimas - that's a good point, this is actually a PR I wrote about 2 months ago, so it can benefit from my new knowledge 😄

**Comment by TimKnight-DWP** [comment]
@amimas - all updated thanks

File: gitlabform/processors/project/job_token_scope_processor.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        )
+        for group_id in group_ids_to_add:
+            job_token_scope.groups_allowlist.create({"target_group_id": group_id})
+            verbose("Added group %s to allowlist", group_id)
+            job_token_scope.save()
+
+    def _add_projects_to_allowlist(
+        self, project, job_token_scope, existing_allowlist, project_ids_to_allow
+    ):
+        project_ids_to_add = self._get_ids_to_add_to_allowlist(
```
Comment (amimas) [inline gitlabform/processors/project/job_token_scope_processor.py:149]: Not sure if it's just the naming of things or if there's some optimization possible... But I find this a little hard to follow. The function already received a list of project IDs to add to allowlist. Here we're calling another function to "get IDs to add to allowlist". I was already a bit confused in the calling function why "existing allowlist" is being passed because I expected this function to just do as the name says... Add some IDs to allowlist. :)

File: gitlabform/processors/project/job_token_scope_processor.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            existing_allowlist, project_ids_to_allow
+        )
+        for project_id in project_ids_to_add:
+            if project_id != project.id:
+                job_token_scope.allowlist.create({"target_project_id": project_id})
+                verbose("Added project %s to allowlist", project_id)
+                job_token_scope.save()
```
Comment (amimas) [inline gitlabform/processors/project/job_token_scope_processor.py:156]: Does it make sense to "save" outside the loop? I assume that way we make a single API call to add all the projects instead of doing it one by one. Maybe that's not how it works in python-gitlab. Not sure.

File: gitlabform/processors/project/job_token_scope_processor.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+        if limit_access_to_this_project != job_token_scope.inbound_enabled:
+            info(
+                f"Updating project job token scope to limit access: {limit_access_to_this_project}"
+            )
+            job_token_scope.enabled = limit_access_to_this_project
+            job_token_scope.save()
+
+            # Refresh object in order for groups_allowlist and allowlist to exist on Scope for modification later
+            job_token_scope.refresh()
```
[the remark this answers is not in the document]
Comment (TimKnight-DWP) [reply gitlabform/processors/project/job_token_scope_processor.py:56]: yeah it appears `refresh()` on python-gitlab objects mutates the object itself, on further thought I think I will re-get it outside the function rather than use refresh/mutation, to avoid confusion like this 👍

[the remark this answers is not in the document]
Comment (TimKnight-DWP) [reply gitlabform/processors/project/job_token_scope_processor.py:56]: also want to test what happens if I disable token_scope _and_ try remove projects/groups from allowlist 🤔

File: gitlabform/processors/project/job_token_scope_processor.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        job_token_scope: ProjectJobTokenScope,
+        groups_allowlist: List,
+        enforce: bool,
+    ):
+        if not groups_allowlist and enforce:
+            warning(
+                "Process will remove existing groups from allowlist, as none set in configuration"
+            )
+
+        existing_allowlist = job_token_scope.groups_allowlist.list()
```
[the remark this answers is not in the document]
Comment (TimKnight-DWP) [reply gitlabform/processors/project/job_token_scope_processor.py:103]: I would _prefer_ to use list() since that matches intended functionality and GitLab API, but I do remember something about an open PR whereby list() can be inconsistent in `python-gitlab` @amimas ?

**Comment by TimKnight-DWP** [comment]
@amimas done - drastically reduced number of API calls and hopefully clarified some of the logic both with reducing complexity of code and comments

File: gitlabform/processors/project/job_token_scope_processor.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                pass
+
+            allowlist_updated = True
+            job_token_scope.groups_allowlist.create({"target_group_id": group_id})
+            verbose("Added Group %s to allowlist", group_id)
+
+        if allowlist_updated:
+            verbose("Saving added Groups allowlist changes")
+            job_token_scope.save()
+            return True
```
Comment (amimas) [inline gitlabform/processors/project/job_token_scope_processor.py:173]: Looks like this function will only return conditionally. I think we should make sure to always return with a boolean type.

**Review by amimas — APPROVED** [review]
Thanks @TimKnight-DWP for all the updates. Found it easy to follow. Have couple of comments that might be applicable for both projects/groups allowlist updating. Let me know please if they make sense. I leave it to you if those updates are needed or not.

Thanks for all the work on implementing this feature. I know it wasn't a small task.

File: gitlabform/processors/project/job_token_scope_processor.py (hunk 6 above)
Comment (TimKnight-DWP) [reply gitlabform/processors/project/job_token_scope_processor.py:173]: rather than relying on truthy vs falsy 👍

File: gitlabform/processors/project/job_token_scope_processor.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    ):
+        allowlist_updated = False
+
+        for group_id in group_ids_listed_in_config:
+            if any(allowed.id == group_id for allowed in existing_allowlist):
+                # If already in allowlist, do nothing
+                verbose(f"{group_id} already in Groups allowlist")
+                pass
+
+            allowlist_updated = True
```
[the remark this answers is not in the document]
Comment (TimKnight-DWP) [reply gitlabform/processors/project/job_token_scope_processor.py:166]: @amimas it will only set to True if there is a group or project to add to the allowlist

If the only groups listed in the config are already on the allowlist, it won't save the scope

I wanted to keep it outside the if statement so we only save once and if necessary.

[the remark this answers is not in the document]
Comment (TimKnight-DWP) [reply gitlabform/processors/project/job_token_scope_processor.py:166]: Very glad you asked this @amimas, I added a test for it and discovered there was a bug in my logic!

**Review by amimas — APPROVED** [review]
@TimKnight-DWP - Unfortunately I'm not able to clearly see the new changes since it was squashed with all prior commits. I skimmed through the main logic again and it seems good to me. It's a big new feature implementation. I'm leaning towards merging this. If needed, we can iterate on this after we find issues. But overall, I think this is ready to ship out. 🚀

**Comment by TimKnight-DWP** [comment]
@amimas - ah yeah sorry should have thought of that when tidying up 😄
